### ISSUE TITLE
[Bug]: error running fastp

### ISSUE THREAD

**Issue by flefler** [franciscozorrilla/metaGEM#138]
Hi Francisco,

I recently downloaded metagem, I used the manual installation guideline. I'm trying to run fastp but keep getting this error. I am unsure what could be causing this. I was able to run createFolders, downloadToy, organizeData, and check with no problems. I am using sample1 from the toy dataset.

`(metagem) dail@swamp:~/metaGEM/workflow$ bash /home/dail/metaGEM/workflow/metaGEM.sh -t fastp -l`

=================================================================================================================================
Developed by: Francisco Zorrilla, Kiran R. Patil, and Aleksej Zelezniak___________________________________________________________
Publication: doi.org/10.1101/2020.12.31.424982___________________________/\\\\\\___/\\\\\\\\___/\\____________/\\_        
________________________________________________________________________/\\//////////___\/\\///////////___\/\\\________/\\\_       
____________________________________________/\\________________________/\\______________\/\\______________\/\\//\\____/\\//\\_      
_______/\\\__/\\\________/\\\\____/\\\\\\___/\\\\\_____\/\\____/\\\\__\/\\\\\\______\/\\///\\/\\/_\/\\_     
______/\\///\\\///\\____/\\/////\\__\////\\////___\////////\\____\/\\___\/////\\__\/\\///////_______\/\\__\///\\/___\/\\_    
______\/\\_\//\\__\/\\___/\\\\\\______\/\\_________/\\\\\___\/\\_______\/\\__\/\\______________\/\\____\///_____\/\\_   
_______\/\\__\/\\__\/\\__\//\///////_______\/\\_/\____/\\/////\\___\/\\_______\/\\__\/\\______________\/\\_____________\/\\_  
________\/\\__\/\\__\/\\___\//\\\\\_____\//\\\____\//\\\\/\__\//\\\\\\/___\/\\\\\\\\__\/\\_____________\/\\_ 
_________\///___\///___\///_____\//////////_______\/////______\////////\//____\////////////_____\///////////////___\///______________\///__
=============================================================================================================================================
A Snakemake-based pipeline desinged to predict metabolic interactions directly from metagenomics data using high performance computer clusters
===============================================================================================================================================
Version: 1.0.5

Setting current directory to root in config.yaml file ... 

Parsing Snakefile to target rule: fastp ... 
 
Do you wish to continue with these parameters? (y/n)y
Proceeding with fastp job(s) ... 

Please verify parameters set in the config.yaml file: 

path:
    root: /home/dail/metaGEM/workflow
    scratch: $TMP
folder:
    data: dataset
    logs: logs
    assemblies: assemblies
    scripts: scripts
    crossMap: crossMap 
    concoct: concoct
    maxbin: maxbin
    metabat: metabat
    refined: refined_bins
    reassembled: reassembled_bins
    classification: GTDBTk
    abundance: abundance
    GRiD: GRiD
    GEMs: GEMs
    SMETANA: SMETANA
    memote: memote
    qfiltered: qfiltered
    stats: stats
    proteinBins: protein_bins
    dnaBins: dna_bins
    pangenome: pangenome
    kallisto: kallisto
    kallistoIndex: kallistoIndex
    benchmarks: benchmarks
    prodigal: prodigal
    blastp: blastp
    blastp_db: blastp_db
scripts:
    kallisto2concoct: kallisto2concoct.py
    prepRoary: prepareRoaryInput.R
    binFilter: binFilter.py
    qfilterVis: qfilterVis.R
    assemblyVis: assemblyVis.R
    binningVis: binningVis.R
    modelVis: modelVis.R
    compositionVis: compositionVis.R
    taxonomyVis: taxonomyVis.R
    carveme: media_db.tsv
    toy: download_toydata.txt
    GTDBtkVis: 
cores:
    fastp: 4
    megahit: 48
    crossMap: 48
    concoct: 48
    metabat: 48
    maxbin: 48
    refine: 48
    reassemble: 48
    classify: 2
    gtdbtk: 48
    abundance: 16
    carveme: 4
    smetana: 12
    memote: 4
    grid: 24
    prokka: 2
    roary: 12
    diamond: 12
params:
    cutfasta: 10000
    assemblyPreset: meta-sensitive
    assemblyMin: 1000
    concoct: 800
    metabatMin: 50000
    seed: 420
    minBin: 1500
    refineMem: 1600
    refineComp: 50
    refineCont: 10
    reassembleMem: 1600
    reassembleComp: 50
    reassembleCont: 10
    carveMedia: M8
    smetanaMedia: M1,M2,M3,M4,M5,M7,M8,M9,M10,M11,M13,M14,M15A,M15B,M16
    smetanaSolver: CPLEX
    roaryI: 90
    roaryCD: 90
envs:
    metagem: envs/metagem
    metawrap: envs/metawrap
    prokkaroary: envs/prokkaroary

Please pay close attention to make sure that your paths are properly configured!
Do you wish to proceed with this config.yaml file? (y/n)y
 
Unlocking snakemake ... 
Unlocking working directory.

Dry-running snakemake jobs ... 
Building DAG of jobs...
Job counts:
	count	jobs
	1	all
	1	qfilter
	2

[Mon Aug 28 20:08:14 2023]
rule qfilter:
    input: /home/dail/metaGEM/workflow/dataset/sample1/sample1_R1.fastq.gz, /home/dail/metaGEM/workflow/dataset/sample1/sample1_R2.fastq.gz
    output: /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz, /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R2.fastq.gz
    jobid: 1
    wildcards: IDs=sample1

[Mon Aug 28 20:08:14 2023]
Job 0: 
        WARNING: Be very careful when adding/removing any lines above this message.
        The metaGEM.sh parser is presently hardcoded to edit line 22 of this Snakefile to expand target rules accordingly,
        therefore adding/removing any lines before this message will likely result in parser malfunction.
        

Job counts:
	count	jobs
	1	all
	1	qfilter
	2
This was a dry-run (flag -n). The order of jobs does not reflect the order of execution.
Do you wish to submit this batch of jobs on your local machine? (y/n)y
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 20
Rules claiming more threads will be scaled down.
Conda environments: ignored
Job counts:
	count	jobs
	1	all
	1	qfilter
	2
Select jobs to execute...

[Mon Aug 28 20:08:16 2023]
rule qfilter:
    input: /home/dail/metaGEM/workflow/dataset/sample1/sample1_R1.fastq.gz, /home/dail/metaGEM/workflow/dataset/sample1/sample1_R2.fastq.gz
    output: /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz, /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R2.fastq.gz
    jobid: 1
    wildcards: IDs=sample1

Activating envs/metagem conda environment ... 
/usr/bin/bash: line 2: activate: No such file or directory
[Mon Aug 28 20:08:16 2023]
Error in rule qfilter:
    jobid: 1
    output: /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz, /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R2.fastq.gz
    shell:
        
        # Activate metagem environment
        echo -e "Activating envs/metagem conda environment ... "
        set +u;source activate envs/metagem;set -u;

        # This is just to make sure that output folder exists
        mkdir -p $(dirname /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz)

        # Make job specific scratch dir
        idvar=$(echo $(basename $(dirname /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz))|sed 's/_R1.fastq.gz//g')
        echo -e "
Creating temporary directory $TMP/qfiltered/${idvar} ... "
        mkdir -p $TMP/qfiltered/${idvar}

        # Move into scratch dir
        cd $TMP/qfiltered/${idvar}

        # Copy files
        echo -e "Copying /home/dail/metaGEM/workflow/dataset/sample1/sample1_R1.fastq.gz and /home/dail/metaGEM/workflow/dataset/sample1/sample1_R2.fastq.gz to $TMP/qfiltered/${idvar} ... "
        cp /home/dail/metaGEM/workflow/dataset/sample1/sample1_R1.fastq.gz /home/dail/metaGEM/workflow/dataset/sample1/sample1_R2.fastq.gz .

        echo -e "Appending .raw to temporary input files to avoid name conflict ... "
        for file in *.gz; do mv -- "$file" "${file}.raw.gz"; done

        # Run fastp
        echo -n "Running fastp ... "
        fastp --thread 4             -i *R1*raw.gz             -I *R2*raw.gz             -o $(basename /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz)             -O $(basename /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R2.fastq.gz)             -j $(dirname /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz)/$(echo $(basename $(dirname /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz))).json             -h $(dirname /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz)/$(echo $(basename $(dirname /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz))).html

        # Move output files to root dir
        echo -e "Moving output files $(basename /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz) and $(basename /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R2.fastq.gz) to $(dirname /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz)"
        mv $(basename /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz) $(basename /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R2.fastq.gz) $(dirname /home/dail/metaGEM/workflow/qfiltered/sample1/sample1_R1.fastq.gz)

        # Warning
        echo -e "Note that you must manually clean up these temporary directories if your scratch directory points to a static location instead of variable with a job specific location ... "

        # Done message
        echo -e "Done quality filtering sample ${idvar}"
        
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Job failed, going on with independent jobs.
Exiting because a job execution failed. Look above for error message
Complete log: /home/dail/metaGEM/workflow/.snakemake/log/2023-08-28T200816.260324.snakemake.log

**Comment by flefler** [comment]
I ran `createFolders` before this. After running `(metagem) dail@swamp:~/metaGEM/workflow$ bash /home/dail/metaGEM/workflow/metaGEM.sh -t fastp -l` the qfiltered folder disappears.

**Comment by franciscozorrilla** [comment]
Hi Forrest,

Thanks for the extensive report, I think I have an idea what might be going on, looks like the important error message is this:

`/usr/bin/bash: line 2: activate: No such file or directory`

Which means you probably have the `metagem` environment set up but maybe its not under the `envs/` subdirectory in the root folder. Please run `conda list` to see the path to your metagem env, and then make sure to replace the default`envs/metagem` field in the `config.yaml` file to point towards the appropriate environment path:

https://github.com/franciscozorrilla/metaGEM/blob/cb099b7a557af8bddbd3b257670d263be3ac5a1f/config/config.yaml#L83-L84

Hope this helps and please let me know if you have further questions.
p.s. if possible, its recommended to run metaGEM on a high performance computer cluster rather than local workstation

Best,
Francisco

**Comment by flefler** [comment]
Hi Francisco,

**Comment by franciscozorrilla** [comment]
Hi Forrest,

Based on your error messages it looks like concoct and maxbin fail before even getting to the binRefine stage. 
It is possible that you may have tmp files from a previous job run getting in the way:

> gzip: contigs.fasta already exists; not overwritten
> gzip: contigs.fasta already exists; not overwritten

Could you try deleting everything inside the tmp folder, and then re-running the jobs? e.g. `rm -r tmp/*`
Also, could you check the log files for those specific concoct/maxbin job runs to see if there is a more specific error message? Normally these would be generated inside your `logs/` folder, but not sure if that is still the case with with local flag usage.

Sorry you are having trouble with this. For some background: the workflow was tested and designed to run on slurm-based HPC infrastructures, with some users even expanding to [qsub-based infrastructures](https://github.com/franciscozorrilla/metaGEM/issues/136). However, since I do not have access to workstations, I have not developed support for this kind of infrastructure. The local command/flag is mostly meant for troubleshooting, debugging, testing, etc. I am happy, however, to help you to the best of my ability to get metaGEM running on your workstation 💎

**Comment by flefler** [comment]
This is the log file 

Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 20
Rules claiming more threads will be scaled down.
Conda environments: ignored
Job counts:
	count	jobs
	1	all
	1	binRefine
	1	concoct
	1	maxbinCross
	1	metabatCross
	5
Select jobs to execute...
Failed to solve scheduling problem with ILP solver. Falling back to greedy solver. Run Snakemake with --verbose to see the full solver output for debugging the problem.

[Wed Aug 30 09:50:53 2023]
rule maxbinCross:
    input: /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz, /home/dail/metaGEM/workflow/maxbin/sample1/cov
    output: /home/dail/metaGEM/workflow/maxbin/sample1/sample1.maxbin-bins
    jobid: 7
    benchmark: /home/dail/metaGEM/workflow/benchmarks/sample1.maxbin.benchmark.txt
    wildcards: IDs=sample1


[Wed Aug 30 09:50:53 2023]
rule concoct:
    input: /home/dail/metaGEM/workflow/concoct/sample1/cov/coverage_table.tsv, /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz
    output: /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins
    jobid: 2
    benchmark: /home/dail/metaGEM/workflow/benchmarks/sample1.concoct.benchmark.txt
    wildcards: IDs=sample1


[Wed Aug 30 09:50:53 2023]
rule metabatCross:
    input: /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz, /home/dail/metaGEM/workflow/metabat/sample1/cov
    output: /home/dail/metaGEM/workflow/metabat/sample1/sample1.metabat-bins
    jobid: 5
    benchmark: /home/dail/metaGEM/workflow/benchmarks/sample1.metabat.benchmark.txt
    wildcards: IDs=sample1

[Wed Aug 30 09:50:53 2023]
Error in rule maxbinCross:
    jobid: 7
    output: /home/dail/metaGEM/workflow/maxbin/sample1/sample1.maxbin-bins
    shell:
        
        # Activate metagem environment
        #set +u;source activate /home/dail/metaGEM/workflow/envs/metagem;set -u;

        # Create output folder
        mkdir -p $(dirname /home/dail/metaGEM/workflow/maxbin/sample1/sample1.maxbin-bins)

        # Make job specific scratch dir
        fsampleID=$(echo $(basename $(dirname /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz)))
        echo -e "
Creating temporary directory /home/dail/metaGEM/workflow/tmp/maxbin/${fsampleID} ... "
        mkdir -p /home/dail/metaGEM/workflow/tmp/maxbin/${fsampleID}

        # Move into scratch dir
        cd /home/dail/metaGEM/workflow/tmp/maxbin/${fsampleID}

        # Copy files to tmp
        cp -r /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz /home/dail/metaGEM/workflow/maxbin/sample1/cov/*.depth .

        echo -e "
Unzipping assembly ... "
        gunzip $(basename /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz)

        echo -e "
Generating list of depth files based on crossMapSeries rule output ... "
        find . -name "*.depth" > abund.list
        
        echo -e "
Running maxbin2 ... "
        run_MaxBin.pl -thread 12 -contig contigs.fasta -out $(basename $(dirname /home/dail/metaGEM/workflow/maxbin/sample1/sample1.maxbin-bins)) -abund_list abund.list
        
        # Clean up un-needed files
        rm *.depth abund.list contigs.fasta

        # Move files into output dir
        mkdir -p $(basename /home/dail/metaGEM/workflow/maxbin/sample1/sample1.maxbin-bins)
        while read bin;do mv $bin $(basename /home/dail/metaGEM/workflow/maxbin/sample1/sample1.maxbin-bins);done< <(ls|grep fasta)
        mv * $(dirname /home/dail/metaGEM/workflow/maxbin/sample1/sample1.maxbin-bins)
        
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Job failed, going on with independent jobs.
[Wed Aug 30 09:50:58 2023]
Error in rule concoct:
    jobid: 2
    output: /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins
    shell:
        
        # Activate metagem environment
        #set +u;source activate /home/dail/metaGEM/workflow/envs/metagem;set -u;

        # Create output folder
        mkdir -p $(dirname /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins)

        # Make job specific scratch dir
        sampleID=$(echo $(basename $(dirname /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz)))
        echo -e "
Creating temporary directory /home/dail/metaGEM/workflow/tmp/concoct/${sampleID} ... "
        mkdir -p /home/dail/metaGEM/workflow/tmp/concoct/${sampleID}

        # Move into scratch dir
        cd /home/dail/metaGEM/workflow/tmp/concoct/${sampleID}

        # Copy files
        cp /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz /home/dail/metaGEM/workflow/concoct/sample1/cov/coverage_table.tsv .

        echo "Unzipping assembly ... "
        gunzip $(basename /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz)

        echo -e "Done. 
Cutting up contigs (default 10kbp chunks) ... "
        cut_up_fasta.py -c 10000 -o 0 -m $(echo $(basename /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz)|sed 's/.gz//') > assembly_c10k.fa
        
        echo -e "
Running CONCOCT ... "
        concoct --coverage_file $(basename /home/dail/metaGEM/workflow/concoct/sample1/cov/coverage_table.tsv)             --composition_file assembly_c10k.fa             -b $(basename $(dirname /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins))             -t 12             -c 800
            
        echo -e "
Merging clustering results into original contigs ... "
        merge_cutup_clustering.py $(basename $(dirname /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins))_clustering_gt1000.csv > $(basename $(dirname /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins))_clustering_merged.csv
        
        echo -e "
Extracting bins ... "
        mkdir -p $(basename /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins)
        extract_fasta_bins.py $(echo $(basename /home/dail/metaGEM/workflow/assemblies/sample1/contigs.fasta.gz)|sed 's/.gz//') $(basename $(dirname /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins))_clustering_merged.csv --output_path $(basename /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins)
        
        # Move final result files to output folder
        mv $(basename /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins) *.txt *.csv $(dirname /home/dail/metaGEM/workflow/concoct/sample1/sample1.concoct-bins)
        
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Job failed, going on with independent jobs.
[Wed Aug 30 09:50:59 2023]
Finished job 5.
1 of 5 steps (20%) done
Exiting because a job execution failed. Look above for error message
Complete log: /home/dail/metaGEM/workflow/.snakemake/log/2023-08-30T095053.149690.snakemake.log

**Comment by franciscozorrilla** [comment]
> Can't locate LWP/Simple.pm in @inc (you may need to install the LWP::Simple module) (@inc contains: /home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/site_perl/5.36/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/site_perl/5.36 /home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/5.36/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/opt/perl/lib/perl5/5.36 /home/linuxbrew/.linuxbrew/lib/perl5/site_perl/5.36/x86_64-linux-thread-multi /home/linuxbrew/.linuxbrew/lib/perl5/site_perl/5.36) at /home/dail/metaGEM/workflow/envs/metagem/bin/run_MaxBin.pl line 4.

Regarding the above maxbin errors, looks like you are missing some perl library. Try could try and install it manually. Looks like others have reported this error message, maybe read throught this issue here: https://github.com/metagenome-atlas/atlas/issues/328

> raise TypeError(
TypeError: Feature names are only supported if all input features have string names, but your input has ['int', 'str'] as feature name / column name types.

Regarding the above concoct errors, looks like it has to do with the sklearn version, see these issues for more details on how to solve this https://github.com/BinPro/CONCOCT/issues/323 https://github.com/BinPro/CONCOCT/issues/322

Regarding metabat2, from the logs it looks like it may have succesfully generated some draft bins.

After fixing the issues with concoct and maxbin then you should be able to run the refinement and reassembly modules. Remember to try cleaning up the tmp dir in between runs to avoid any existing intermediate file issues that may cause problems.
Let me know if this helps.

Best,
Francisco

**Comment by wupeng1998** [comment]
Hi, I'm using pbs to submit jobs to the server, but as with the command line alone, I have to use the "--local" flag, otherwise I get "This was a dry-run (flag -n). The order of jobs does not reflect the order of execution." and no program will run.
Here is my PBS command, I have to specify the "--local" flag:
#PBS -N draft_bin
#PBS -o draft_bin.log
#PBS -j oe
#PBS -l walltime=10000:00:00
#PBS -l nodes=1:ppn=48
#PBS -l mem=500gb
#PBS -q high

projectname="metaGEM_comand"
project="/public/home/wangjj/WP/metaGEM/workflow"

cd $project

echo -e ""
echo -e ""
source activate metagem

yes | bash metaGEM.sh -t fastp -j 2 -c 120 -m 500 -h 24 --local
echo -e "************************************************************************************************"
echo -e "************************ fastp Finish...!!! ************************"
echo -e "************************************************************************************************"
yes | bash metaGEM.sh -t megahit -j 2 -c 120 -m 500 -h 24 --local
echo -e "************************************************************************************************"
echo -e "************************ megahit Finish...!!! ************************"
echo -e "************************************************************************************************"
yes | bash metaGEM.sh -t crossMapSeries -j 2 -c 120 -m 500 -h 24 --local
echo -e "************************************************************************************************"
echo -e "************************ crossMapSeries Finish...!!! ************************"
echo -e "************************************************************************************************"
yes | bash metaGEM.sh -t concoct -j 2 -c 120 -m 500 -h 24 --local

**Comment by wupeng1998** [comment]
In addition to that, I also found that when using "--local", it seems that no matter how many nodes are applied to the server, the program will only run on one node, and not multiple nodes at the same time in parallel, because I got an out of memory error when I ran GTDB-TK, but in fact, I applied 10 24-core 96GB nodes when I ran GTDB-TK with "--local", and it shouldn't have generated this error:
[2023-10-17 15:35:05] INFO: Masked bacterial alignment from 41,084 to 5,037 AAs.
[2023-10-17 15:35:05] INFO: 0 bacterial user genomes have amino acids in <10.0% of columns in filtered MSA.
[2023-10-17 15:35:05] INFO: Creating concatenated alignment for 45,560 bacterial GTDB and user genomes.
[2023-10-17 15:35:06] INFO: Creating concatenated alignment for 5 bacterial user genomes.
[2023-10-17 15:35:06] INFO: Done.
[2023-10-17 15:35:06] WARNING: pplacer requires ~215 GB of RAM to fully load the bacterial tree into memory. However, 65.44 GB was detected. This may affect pplacer performance, or fail if there is insufficient swap space.
[2023-10-17 15:35:06] TASK: Placing 5 bacterial genomes into reference tree with pplacer using 48 CPUs (be patient).
[2023-10-17 15:35:06] INFO: pplacer version: v1.1.alpha19-0-g807f6f3

**Comment by wupeng1998** [comment]
I think it might be a problem with the conditional judgment here, but I don't know why I get this message when I don't set the "--local" flag: "This was a dry-run (flag -n). The order of jobs does not reflect the order of execution."

 In，metagem.sh
 # Parse snakefile for cluster/local jobs
  elif [ $task == "fastp" ]; then
    string='expand(config["path"]["root"]+"/"+config["folder"]["qfiltered"]+"/{IDs}/{IDs}_R1.fastq.gz", IDs = IDs)'
    if [ $local == "true" ]; then
        submitLocal
    else
        submitCluster
    fi

**Comment by franciscozorrilla** [comment]
Hey wupeng,

>  but as with the command line alone, I have to use the "--local" flag

I do not understand the reasoning here. Why are you trying to use the local flag when submitting jobs to the cluster? In general, you should never run jobs locally on the cluster. The local flag is only for usage with workstations, where there is no scheduler or other users.

> In addition to that, I also found that when using "--local", it seems that no matter how many nodes are applied to the server, the program will only run on one node, and not multiple nodes at the same time in parallel, because I got an out of memory error when I ran GTDB-TK, but in fact, I applied 10 24-core 96GB nodes when I ran GTDB-TK with "--local", and it shouldn't have generated this error:

This behavior is exactly as expected. To submit jobs to the cluster, instead of running them locally, remove the --local flag. In fact, if you add the local flag, Snakemake will launch the jobs directly in the node you are running and that is why you are running out of memory.

> I'm using pbs to submit jobs to the server

If you are using a pbs cluster then have a look a this [discussion](https://github.com/franciscozorrilla/metaGEM/issues/136) and this [fork](https://github.com/fbartusch/metaGEM/tree/master). @fbartusch modified the metaGEM.sh wrapper file and cluster config file to allow submission on qsub, if I were you I would look at those modifications and apply them to your files as well

Hope this helps and let me know if you have further questions!
Best,
Francisco

p.s. feel free to open up a new issue/discussion :)

**Comment by wupeng1998** [comment]
> Hey Wupeng,
> 
> > but as with the command line alone, I have to use the "--local" flag
> 
> I do not understand the reasoning here. Why are you trying to use the local flag when submitting jobs to the cluster? In general, you should never run jobs locally on the cluster. The local flag is only for usage with workstations, where there is no scheduler or other users.
> 
> > In addition to that, I also found that when using "--local", it seems that no matter how many nodes are applied to the server, the program will only run on one node, and not multiple nodes at the same time in parallel, because I got an out of memory error when I ran GTDB-TK, but in fact, I applied 10 24-core 96GB nodes when I ran GTDB-TK with "--local", and it shouldn't have generated this error:
> 
> This behavior is exactly as expected. To submit jobs to the cluster, instead of running them locally, remove the --local flag. In fact, if you add the local flag, Snakemake will launch the jobs directly in the node you are running and that is why you are running out of memory.
> 
> > I'm using pbs to submit jobs to the server
> 
> If you are using a pbs cluster then have a look a this [discussion](https://github.com/franciscozorrilla/metaGEM/issues/136) and this [fork](https://github.com/fbartusch/metaGEM/tree/master). @fbartusch modified the metaGEM.sh wrapper file and cluster config file to allow submission on qsub, if I were you I would look at those modifications and apply them to your files as well
> 
> Hope this helps and let me know if you have further questions! Best, Francisco
> 
> p.s. feel free to open up a new issue/discussion :)

Thanks for your reply, I did find this problem too, when I unchecked "--local", the nohup.out file mentions that the "sbatch" command can't be found, so as you mentioned, it's indeed a problem with qsub! I will take the next step as you suggested, thanks again!

**Comment by franciscozorrilla** [comment]
Hi wupeng, it sounds like your issues are no longer related to the original post. Please feel free to open up a new issue and provide more details. Based on the error message, it seems like Snakemake is not properly communicating with your cluster. First make sure that Snakemake is working properly, you should be able to submit/run simple Snakemake jobs on your cluster before trying to use metaGEM. For reference, please see the [Snakemake docs and tutorial](https://snakemake.readthedocs.io/en/stable/tutorial/tutorial.html).

**Comment by wupeng1998** [comment]
Ok,i know what you means. But the snakemake is work well when i use the --local flag.so i think there maybe not the snakemake false

**Comment by franciscozorrilla** [comment]
Please, do not run jobs on the login node of your cluster using the --local flag. This is improper and harmful usage of the cluster, and you will probably get complaints from your HPC admins. Snakemake is supposed to communicate with your HPC job scheduler and submit jobs to compute nodes, as opposed to launching them on the login node which is what you are doing with the --local flag.
